modules/cell-type-ewings/resources/usr/bin/assign-celltypes.R

```diff
@@ -0,0 +1,151 @@
+    help = "Path to file where results will be saved"
+  )
+)
+
+# Parse options
+opt <- parse_args(OptionParser(option_list = option_list))
+
+# Set up -----------------------------------------------------------------------
+
+# make sure input files exist
+stopifnot(
+  "sce file does not exist" = file.exists(opt$sce_file),
+  "consensus cell types file does not exist" = file.exists(opt$consensus_celltype_file),
+  "AUCell results file does not exist" = file.exists(opt$aucell_results_file),
+  "auc_table file does not exist" = file.exists(opt$auc_table),
+  "cell_state_markers_file does not exist" = file.exists(opt$cell_state_markers_file)
+)
+
+# load SCE
+suppressPackageStartupMessages({
+  library(SingleCellExperiment)
+})
+
+# read in files
+sce <- readr::read_rds(opt$sce_file)
+consensus_df <- readr::read_tsv(opt$consensus_celltype_file)
+aucell_results_df <- readr::read_tsv(opt$aucell_results_file)
+auc_threshold_df <- readr::read_tsv(opt$auc_table)
+cell_state_markers_df <- readr::read_tsv(opt$cell_state_markers_file)
+
+
+# assign normal cell types using consensus cell types --------------------------
+assignment_df <- consensus_df |>
+  dplyr::mutate(
+    custom_annotation = dplyr::if_else(
+      !is.na(consensus_ontology),
+      consensus_annotation,
+      "unknown"
+    )
+  )
+
+
+# assign tumor cells and tumor cell states -------------------------------------
+# pull out tumor cell states
+tumor_cell_states <- unique(auc_threshold_df$cell_type)
+
+# for each cell state get a list of cells that belong to that cell state using AUCell results
+# then update the existing custom annotation column to label those cells with that cell state
+tumor_cell_states |>
+  purrr::walk(\(state){
+    # get a list of all gene sets that should be used to classify that cell state
+    geneset_auc_list <- auc_threshold_df |>
+      dplyr::filter(cell_type == state) |>
+      dplyr::select(gene_set, auc_threshold) |>
+      tibble::deframe()
+
+    # get cells that meet all criteria, e.g., have auc > threshold for all gene sets
+    cells <- geneset_auc_list |>
+      purrr::imap(\(auc_threshold, name){
+        aucell_results_df |>
+          dplyr::filter(gene_set == name & auc > auc_threshold) |>
+          dplyr::pull(barcodes)
+      }) |>
+      purrr::reduce(intersect)
+
+    # figure out which cells these are
+    cell_state_idx <- which(assignment_df$barcodes %in% cells)
+
+    # label the cells with that cell state
+    # modify the data frame in place with <<-
+    assignment_df$custom_annotation[cell_state_idx] <<- state
+  })
+
+# Label proliferative tumor cells ----------------------------------------------
+
+# get proliferative marker genes
+proliferative_markers <- cell_state_markers_df |>
+  dplyr::filter(cell_state == "proliferative") |>
+  dplyr::pull(ensembl_gene_id)
+
+# calculate the mean expression of all genes for all cells
+mean_exp <- logcounts(sce[proliferative_markers, ]) |>
+  as.matrix() |>
+  t() |>
+  rowMeans()
```

Comment:
This is the only place where you use the SCE in this script. Would it be possible to have this be calculated during as part of the AUCell script or process? It might be nice not to have to pass the SCE files into this process, if we can avoid it.

Reply:
Separately, any reason this doesn't work? 
```suggestion
mean_exp <- logcounts(sce)[proliferative_markers, ] |>
  colMeans()
```

This is a "works for me" situation, mostly, but I also want to note that I also pulled the row selection to after getting the log counts, to save subsetting the whole object.

Reply:
Yeah, I think it would have to be a separate file; I still think that is preferable to passing around the whole object.

Reply:
How do you feel about making a second script that calculates the mean expression for the all the custom gene sets in https://github.com/AlexsLemonade/OpenScPCA-analysis/blob/main/analyses/cell-type-ewings/references/tumor-cell-state-markers.tsv, not just proliferative. Then we can run that script as a secondary step within the same process that runs AUCell?

Reply:
That works for me. While there would be an efficiency gain from not loading the whole object in two different scripts, the modularity is probably worth the loss.